pkg/app/appconf/store.go

```diff
@@ -0,0 +1,298 @@
+package appconf
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/google/renameio/v2/maybe"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+
+	"github.com/vanti-dev/sc-bos/internal/confmerge"
+	"github.com/vanti-dev/sc-bos/pkg/auto"
+	"github.com/vanti-dev/sc-bos/pkg/block"
+	"github.com/vanti-dev/sc-bos/pkg/driver"
+	"github.com/vanti-dev/sc-bos/pkg/zone"
+)
+
+type Store struct {
+	logger  *zap.Logger
+	backing confmerge.Store
+
+	m      sync.Mutex
+	active Config
+}
+
+func LoadStore(external Config, schema Schema, storeDir string, logger *zap.Logger) (*Store, error) {
+	store := confmerge.NewDirStore(storeDir)
+	active, patches, err := confmerge.Merge(external, store, schema.Blocks())
+	err = multierr.Append(err, saveConfigPatches(patches, filepath.Join(storeDir, patchDirName), logger))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Store{
+		logger:  logger,
+		backing: store,
+		active:  active,
+	}, nil
+}
+
+func (s *Store) Active() Config {
+	s.m.Lock()
+	defer s.m.Unlock()
+	return s.active.clone()
+}
+
+func (s *Store) Drivers() *DriverStore {
+	return &DriverStore{store: s}
+}
+
+func (s *Store) Automations() *AutomationStore {
+	return &AutomationStore{store: s}
+}
+
+func (s *Store) Zones() *ZoneStore {
+	return &ZoneStore{store: s}
+}
+
+func (s *Store) save(updated Config) error {
+	encoded, err := json.MarshalIndent(updated, "", "  ")
+	if err != nil {
+		return err
+	}
+	err = s.backing.SetActiveConfig(encoded)
+	if err != nil {
+		return err
+	}
+	s.active = updated
+	return nil
+}
+
+type DriverStore struct {
+	store *Store
+}
+
+func (ds *DriverStore) SaveConfig(_ context.Context, name string, typ string, data []byte) error {
```

Comment:
It is very annoying that these type stores can't be more generic

Reply:
I'll have a go at it, but my gut feeling is that it won't reduce the line count much